txpool/txpool.go

```diff
@@ -120,13 +122,76 @@ func NewTxPool(
 	return txPool, nil
 }
 
+// accountQueueWrapper is the account based queue mux map implementation
+type accountQueueWrapper struct {
+	mux          sync.RWMutex // mux for accessing the accountQueue
+	writeLock    int32        // flag indicating whether a write lock is held
+	accountQueue *txHeapWrapper
+}
+
+// lockAccountQueue returns the corresponding account queue wrapper object, or creates it
+// if it doesn't exist in the account queue map
+func (t *TxPool) lockAccountQueue(address types.Address, writer bool) *accountQueueWrapper {
+	// Lock the global map
+	t.accountQueuesMux.Lock()
+
+	accountQueue, ok := t.accountQueues[address]
+	if !ok {
+		// Account queue is not initialized yet, initialize it
+		stateRoot := t.store.Header().StateRoot
+
+		// Initialize the account based transaction heap
+		txnsQueue := newTxHeapWrapper()
+		txnsQueue.nextNonce = t.store.GetNonce(stateRoot, address)
+
+		accountQueue = &accountQueueWrapper{accountQueue: txnsQueue}
+		t.accountQueues[address] = accountQueue
+	}
+	// Unlock the global map, since work is finished
+	t.accountQueuesMux.Unlock()
+
+	// Grab the lock for the specific account queue
+	if writer {
+		accountQueue.mux.Lock()
+		atomic.StoreInt32(&accountQueue.writeLock, 1)
+	} else {
+		accountQueue.mux.RLock()
+		atomic.StoreInt32(&accountQueue.writeLock, 0)
+	}
+
+	return accountQueue
+}
+
+// unlock releases the account specific transaction queue mux.
+// Separated out into a function in case there needs to be additional teardown logic.
+// Code calling unlock shouldn't need to know the type of lock for the mux (writer / reader) to unlock it
+func (a *accountQueueWrapper) unlock() {
+	// Grab the previous lock type and reset it
+	if atomic.SwapInt32(&a.writeLock, 0) == 1 {
+		a.mux.Unlock()
+	} else {
+		a.mux.RUnlock()
+	}
+}
+
```

Comment:
What would happen in following case?
Process1: call `lockAccountQueue` as writer
Process2: call `lockAccountQueue` as reader
Process1: call `unlock` => what happens?

Reply:
> What would happen in following case? Process1: call `lockAccountQueue` as writer Process2: call `lockAccountQueue` as reader Process1: call `unlock` => what happens?

The writer / reader would block up until the other process gives back control of the lock, but **before that happens, it resets the atomic value indicating the type of lock**, so I don't think it's a race condition.

> I think you might want to let accountQueue have RWMutex. So you need to lock twice.
> 
>     1. Get account queue from `accountQueues`
> 
>     2. Read/Write any data in `accountQueue`

What do you mean by this?